lib/Doctrine/MongoDB/Collection.php

```diff
@@ -1277,10 +1297,25 @@ protected function doGroup($keys, array $initial, $reduce, array $options)
     protected function doInsert(array &$a, array $options)
     {
         $document = $a;
+
         $options = isset($options['safe']) ? $this->convertWriteConcern($options) : $options;
         $options = isset($options['wtimeout']) ? $this->convertWriteTimeout($options) : $options;
         $options = isset($options['timeout']) ? $this->convertSocketTimeout($options) : $options;
-        $result = $this->mongoCollection->insert($document, $options);
+
+        $insert = function () use (&$document, $options) {
+            $insert = $document;
```

Comment:
This very confusing chain of referencing and value passing is to allow the `retryInsert` function to generate an id and insert it into the document and allow it to be visible when performing the underlying insertion command as well as in the return value.